src/GraphQL/GraphQLBuilderExtensions.cs

```diff
@@ -0,0 +1,827 @@
+        {
+            builder.Register<IComplexityAnalyzer>(analyzer ?? throw new ArgumentNullException(nameof(analyzer)));
+            builder.ConfigureExecution(opts =>
+            {
+                opts.ComplexityConfiguration ??= new ComplexityConfiguration();
+                action?.Invoke(opts.ComplexityConfiguration, opts.RequestServices);
+            });
+            return builder;
+        }
+
+        /// <summary>
+        /// Registers a singleton of type <see cref="IComplexityAnalyzer"/> within the dependency injection framework
+        /// using the specified factory delegate, then enables and configures it with the specified configuration delegate.
+        /// </summary>
+        public static IGraphQLBuilder AddComplexityAnalyzer<TAnalyzer>(this IGraphQLBuilder builder, Func<IServiceProvider, TAnalyzer> analyzerFactory, Action<ComplexityConfiguration>? action = null)
+            where TAnalyzer : class, IComplexityAnalyzer
+        {
+            builder.Register<IComplexityAnalyzer>(analyzerFactory ?? throw new ArgumentNullException(nameof(analyzerFactory)), ServiceLifetime.Singleton);
+            builder.ConfigureExecution(opts =>
+            {
+                opts.ComplexityConfiguration ??= new ComplexityConfiguration();
+                action?.Invoke(opts.ComplexityConfiguration);
+            });
+            return builder;
+        }
+
+        /// <inheritdoc cref="AddComplexityAnalyzer{TAnalyzer}(IGraphQLBuilder, Func{IServiceProvider, TAnalyzer}, Action{ComplexityConfiguration})"/>
+        public static IGraphQLBuilder AddComplexityAnalyzer<TAnalyzer>(this IGraphQLBuilder builder, Func<IServiceProvider, TAnalyzer> analyzerFactory, Action<ComplexityConfiguration, IServiceProvider?>? action)
+            where TAnalyzer : class, IComplexityAnalyzer
+        {
+            builder.Register<IComplexityAnalyzer>(analyzerFactory ?? throw new ArgumentNullException(nameof(analyzerFactory)), ServiceLifetime.Singleton);
+            builder.ConfigureExecution(opts =>
+            {
+                opts.ComplexityConfiguration ??= new ComplexityConfiguration();
+                action?.Invoke(opts.ComplexityConfiguration, opts.RequestServices);
+            });
+            return builder;
+        }
+        #endregion
+
+        #region - AddErrorInfoProvider
+        /// <summary>
+        /// Configures the default error info provider with the specified configuration delegate.
+        /// </summary>
+        public static IGraphQLBuilder AddErrorInfoProvider(this IGraphQLBuilder builder, Action<ErrorInfoProviderOptions>? action = null)
+            => builder.AddErrorInfoProvider<ErrorInfoProvider>().Configure(action);
+
+        /// <inheritdoc cref="AddErrorInfoProvider(IGraphQLBuilder, Action{ErrorInfoProviderOptions})"/>
+        public static IGraphQLBuilder AddErrorInfoProvider(this IGraphQLBuilder builder, Action<ErrorInfoProviderOptions, IServiceProvider> action)
```

Comment:
I don't see `= null` here for `action`. You talked that it's convinient to have `= null`.

Reply:
You're looking at the wrong overload.  If both have `= null` then the compiler can't decide which one to use if calling `.AddErrorInfoProvider()`.  So the 'first' overload will always have an `= null` while the second won't.

```cs
//overload #1
public static IGraphQLBuilder AddErrorInfoProvider(this IGraphQLBuilder builder, Action<ErrorInfoProviderOptions>? action = null);

//overload #2
public static IGraphQLBuilder AddErrorInfoProvider(this IGraphQLBuilder builder, Action<ErrorInfoProviderOptions, IServiceProvider> action);
```

Curiously, I don't see the `?` on the `action` argument of the second overload.  Perhaps I should check and be sure I was consistent with the NRT markings throughout.

Reply:
This is the only `?` that I am missing.  All the rest are consistent.  #2785

Reply:
OK